### PR TITLE
PMM-10270 PMM-12191 restore-mongo-backup-on-new-service.

### PR DESCRIPTION
**Pull request by PavelKhripkov**
PMM-10270
PMM-12191

Link to the Feature Build: SUBMODULES-3295

### CODE REVIEW

File: agent/runner/jobs/pbm_helpers.go [hunk 1]
```diff
@@ -557,21 +550,49 @@ func groupPartlyDoneErrors(info describeInfo) error {
...
+		select {
+		case <-ticker.C:
+			checks++
+			status, err := getPBMStatus(ctx, dbURL)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(status.Backups.Snapshot) == 0 {
+				l.Debugf("Try number %d of getting list of artifacts from PBM is failed.", checks)
```
Comment (atymchuk) [inline agent/runner/jobs/pbm_helpers.go:585]:
```suggestion
				l.Debugf("Attempt %d to get a list of PBM artifacts has failed.", checks)
```

File: managed/models/artifact_helpers.go [hunk 2]
```diff
@@ -67,6 +69,13 @@ func FindArtifacts(q *reform.Querier, filters ArtifactFilters) ([]*Artifact, err
 	if filters.Status != "" {
 		conditions = append(conditions, fmt.Sprintf("status = %s", q.Placeholder(idx)))
 		args = append(args, filters.Status)
+		idx++
+	}
+
+	if filters.Folder != nil {
+		conditions = append(conditions, fmt.Sprintf("folder = %s", q.Placeholder(idx)))
+		args = append(args, *filters.Folder)
+		// idx++
```
Comment (atymchuk) [inline managed/models/artifact_helpers.go:78]: Leftover?

File: managed/services/preconditions.go [hunk 3]
```diff
@@ -87,3 +89,85 @@ func CheckMongoDBBackupPreconditions(q *reform.Querier, mode models.BackupMode,
 
 	return nil
 }
+
+// CheckArtifactOverlapping checks if there are other artifacts or scheduled tasks pointing to the same location and folder.
+// Placing MySQL and MongoDB artifacts in the same folder is not desirable, while placing MongoDB artifacts of different clusters
+// in the same folder may cause data inconsistency.
+//
+// WARNING: This function valid only when executed as part of transaction with serializable isolation level.
```
Comment (atymchuk) [inline managed/services/preconditions.go:97]:
```suggestion
// WARNING: This function is valid only when executed as part of transaction with serializable isolation level.
```

File: managed/services/preconditions.go [hunk 4]
```diff
@@ -87,3 +89,85 @@ func CheckMongoDBBackupPreconditions(q *reform.Querier, mode models.BackupMode,
...
+func CheckArtifactOverlapping(q *reform.Querier, serviceID, locationID, folder string) error {
+	// TODO This doesn't work for all cases. For example, there are may exist more than one storage locations pointing to the same place.
```
Comment (atymchuk) [inline managed/services/preconditions.go:99]:
```suggestion
	// TODO This doesn't work for all cases. For example, there may exist more than one storage locations pointing to the same place.
```

File: managed/services/preconditions.go [hunk 5]
```diff
@@ -87,3 +89,85 @@ func CheckMongoDBBackupPreconditions(q *reform.Querier, mode models.BackupMode,
...
+			if service.ServiceType == models.MongoDBServiceType && svc.ServiceType == models.MongoDBServiceType {
+				if svc.Cluster != service.Cluster {
+					return errors.Wrapf(ErrLocationFolderPairAlreadyUsed,
+						"Same location and folder already used for artifact %s of other service: %s", artifact.ID, serviceID)
+				}
+				continue
+			}
+
+			return errors.Wrapf(ErrLocationFolderPairAlreadyUsed,
+				"Same location and folder already used for artifact %s of other service: %s", artifact.ID, serviceID)
```
Comment (atymchuk) [inline managed/services/preconditions.go:134]: Seems to be the same as line 128. Maybe need a const?

**Review by atymchuk — COMMENTED** [review]
Mostly nits, but otherwise great stuff!

File: managed/models/artifact_helpers.go (hunk 2 above)
Comment (PavelKhripkov) [reply managed/models/artifact_helpers.go:78]: Nope, I leave it to not forget uncomment it if adding more fields to filter.

File: managed/services/preconditions.go (hunk 5 above)
Comment (PavelKhripkov) [reply managed/services/preconditions.go]: Done.

File: agent/runner/jobs/pbm_helpers.go (hunk 1 above)
Comment (PavelKhripkov) [reply agent/runner/jobs/pbm_helpers.go]: Done.

File: managed/services/preconditions.go (hunk 3 above)
Comment (PavelKhripkov) [reply managed/services/preconditions.go]: Done.

File: managed/services/preconditions.go (hunk 4 above)
Comment (PavelKhripkov) [reply managed/services/preconditions.go]: Done.

File: agent/runner/jobs/mongodb_restore_job.go [hunk 6]
```diff
@@ -172,44 +177,25 @@ func (j *MongoDBRestoreJob) Run(ctx context.Context, send Send) error {
...
+		return nil, err
+	}
 
-			// Old artifacts don't contain pbm backup name.
-			if snapshotName == "" {
-				return &list.Snapshots[len(list.Snapshots)-1], nil
-			}
+	// Old artifacts don't contain pbm backup name.
+	if snapshotName == "" {
+		return &snapshots[0], nil
```
Comment (PavelKhripkov) [inline agent/runner/jobs/mongodb_restore_job.go:190]: We changed command to get snapshots, and that new one returns snapshots in opposite order.

**Review by artemgavrilov — APPROVED** [review]

**Review by atymchuk — APPROVED** [review]

**Pull request merged**